File: packages/core/src/types/prices.ts
import * as z from 'zod'
import { Product, productSchema } from './products'

const priceIdSchema = z.string().trim().min(1)
const productIdSchema = z.string().trim().min(1)
const unixTimestampSchema = z.number().int().nonnegative()

export const priceSchema: z.ZodType<Price> = z.strictObject({
  object: z.literal('price'),
  id: priceIdSchema,
  productId: productIdSchema,
  unitAmount: z.number().int().nullable(),
  currency: z.string(),
  billingInterval: z.enum(['month', 'year']).nullable(),
  intervalCount: z.number().int().nullable(),
  status: z.string(),
  active: z.boolean(),
  lookupKey: z.string().nullable(),
  nickname: z.string().nullable(),
  type: z.string(),
  billingScheme: z.string(),
  tiersMode: z.string().nullable(),
  tiers: z.record(z.string(), z.unknown()).nullable(),
  recurring: z.record(z.string(), z.unknown()).nullable(),
  taxBehavior: z.string().nullable(),
  transformQuantity: z.record(z.string(), z.unknown()).nullable(),
  unitAmountDecimal: z.string().nullable(),
  trialPeriodDays: z.number().int().nullable(),
  metadata: z.record(z.string(), z.unknown()).nullable(),
  archivedAt: unixTimestampSchema.nullable(),
  product: z.lazy(() => productSchema).nullable(),
  createdAt: unixTimestampSchema,
  updatedAt: unixTimestampSchema,
})

export type Price = {
  object: 'price'
  id: string
  productId: string
  unitAmount: number | null
  currency: string
  billingInterval: 'month' | 'year' | null
  intervalCount: number | null
  status: string
  active: boolean
  lookupKey: string | null
  nickname: string | null
  type: string
  billingScheme: string
  tiersMode: string | null
  tiers: Record<string, unknown> | null
  recurring: Record<string, unknown> | null
  taxBehavior: string | null
  transformQuantity: Record<string, unknown> | null
  unitAmountDecimal: string | null
  trialPeriodDays: number | null
  metadata: Record<string, unknown> | null
  archivedAt: number | null
  product: Product | null
  createdAt: number
}

export type PriceCreateParams = {
  productId: string
  unitAmount?: number | null
  currency: string
  type: string
  billingScheme: string
  nickname?: string | null
  lookupKey?: string | null
  active?: boolean
  metadata?: Record<string, unknown> | null
}

export type PriceUpdateParams = Partial<
  Omit<PriceCreateParams, 'productId' | 'currency' | 'type' | 'billingScheme'>
>

export type PriceListParams = {
  productId?: string
  limit?: number
  startingAfter?: string
  endingBefore?: string
}

export const priceCreateParamsSchema = z.strictObject({
  productId: productIdSchema,
  unitAmount: z.number().int().nullable().optional(),
  currency: z.string().length(3),
  type: z.string(),
  billingScheme: z.string(),
  nickname: z.string().nullable().optional(),
  lookupKey: z.string().nullable().optional(),
  active: z.boolean().optional(),
  metadata: z.record(z.string(), z.unknown()).nullable().optional(),
}) satisfies z.ZodType<PriceCreateParams>

export const priceUpdateParamsSchema = priceCreateParamsSchema.partial().omit({
  productId: true,
  currency: true,
  type: true,
  billingScheme: true,
}) satisfies z.ZodType<PriceUpdateParams>

export const priceListParamsSchema = z.strictObject({
  productId: productIdSchema.optional(),
  limit: z.number().int().min(1).max(100).optional(),
  startingAfter: priceIdSchema.optional(),
  endingBefore: priceIdSchema.optional(),
}) satisfies z.ZodType<PriceListParams>

File: apps/api/db/models/prices.py
from typing import TYPE_CHECKING, Any

from sqlalchemy import (
    JSON,
    BigInteger,
    Boolean,
    ForeignKey,
    Integer,
    String,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from db.models.base import Base

if TYPE_CHECKING:
    from db.models.products import Product
    from db.models.subscription_items import SubscriptionItem


class Price(Base):
    """A specific price point on a ``Product``, mirroring Stripe's ``Price``."""

    __tablename__ = "prices"

    id: Mapped[str] = mapped_column(String, primary_key=True)
    product_id: Mapped[str] = mapped_column(
        String, ForeignKey("products.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # Legacy - Do not remove until migration is complete
    billing_interval: Mapped[str | None] = mapped_column(String, nullable=True)
    interval_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    # Legacy - stripe_price_id removed
    status: Mapped[str] = mapped_column(String, nullable=False, server_default="active")

    # Modified/New Phase 1 fields
    unit_amount: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    unit_amount_decimal: Mapped[str | None] = mapped_column(String, nullable=True)
    currency: Mapped[str] = mapped_column(String(3), nullable=False, server_default="jmd")

    lookup_key: Mapped[str | None] = mapped_column(String, unique=True, nullable=True)
    name: Mapped[str | None] = mapped_column(String, nullable=True)
    nickname: Mapped[str | None] = mapped_column(String, nullable=True)
    type: Mapped[str] = mapped_column(String, nullable=False, server_default="recurring") # one_time, recurring
    billing_scheme: Mapped[str] = mapped_column(String, nullable=False, server_default="per_unit") # per_unit, tiered
    tiers_mode: Mapped[str | None] = mapped_column(String, nullable=True) # graduated, volume
    tiers: Mapped[dict[str, Any] | None] = mapped_column(JSON, nullable=True)
    recurring: Mapped[dict[str, Any] | None] = mapped_column(JSON, nullable=True)
    tax_behavior: Mapped[str | None] = mapped_column(String, nullable=True)
    transform_quantity: Mapped[dict[str, Any] | None] = mapped_column(JSON, nullable=True)
    trial_period_days: Mapped[int | None] = mapped_column(Integer, nullable=True)
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")
    metadata_: Mapped[dict[str, Any] | None] = mapped_column("metadata", JSON, nullable=True)

    created_at: Mapped[int] = mapped_column(BigInteger, nullable=False)
    updated_at: Mapped[int] = mapped_column(BigInteger, nullable=False)
    archived_at: Mapped[int | None] = mapped_column(BigInteger, nullable=True)

    product: Mapped["Product"] = relationship("Product", back_populates="prices")
    subscription_items: Mapped[list["SubscriptionItem"]] = relationship(
        "SubscriptionItem", back_populates="price"
    )
